Index: lib/common/tango/core/Thread.d
===================================================================
--- lib/common/tango/core/Thread.d	(revision 4298)
+++ lib/common/tango/core/Thread.d	(working copy)
@@ -3340,6 +3340,7 @@
         }
         else version( LLVM_AsmX86_Posix )
         {
+            push( 0x00000000 );
             push( cast(size_t) &fiber_entryPoint );                 // EIP
             push( 0x00000000 );                                     // newp
             push( 0x00000000 );                                     // oldp
